### PR TITLE
Timekeeper

### PR DESCRIPTION
**Pull request by vogievetsky**
Are you ready to read the word `timekeeper` a lot?

Code:
- Removed dataCube.maxTime and moved it into Timekeeper the idea is that Timekeeper is in charge of all time keeping and serves as a cache for dataCube max times also. This is needed to ensure that the settings are not constantly updated to keep up with the passage of time.
- a `timekeeper` object is threaded into any function that used to implicitly reply on `new Date()`
- Added `TimeMonitor` - the thing on the server that periodically pols dataCubes for updates.
- Removed `refresh` from `refreshRule`. Might want to add `DataCube#defaultAutoRefresh` as a parameter

Build:
- Playing with native TS build and lint
- Mocks are now auto generated by `mkmodel`

### CODE REVIEW

File: src/common/utils/time-monitor/time-monitor.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { Timekeeper } from "../../models/timekeeper/timekeeper";
+
+export interface Check {
+  (): Q.Promise<Date>;
+}
+
+export class TimeMonitor {
+  public logger: Logger;
+  public regularCheckInterval: number;
+  public specialCheckInterval: number;
```
Comment (evasomething) [inline src/common/utils/time-monitor/time-monitor.ts]: what is this for?

File: src/common/models/filter-clause/filter-clause.ts [hunk 2]
```diff
@@ -75,7 +75,7 @@ export class FilterClause implements Instance<FilterClauseValue, FilterClauseJS>
 
   static evaluate(selection: Expression, now: Date, maxTime: Date, timezone: Timezone): TimeRange {
     if (!selection) return null;
-    var maxTimeMinuteTop = minute.shift(minute.floor(maxTime || now, timezone), timezone, 1);
+    var maxTimeMinuteTop = minute.ceil(maxTime || now, timezone);
```
Comment (evasomething) [inline src/common/models/filter-clause/filter-clause.ts]: these aren't the same right? like shifting a floored value by 1 could return something different from ceiling it?

File: src/client/pivot-entry.ts [hunk 3]
```diff
@@ -57,11 +57,11 @@ require.ensure([
...
-  const WallTime = require('chronoshift').WallTime;
-  const Ajax = require('./utils/ajax/ajax').Ajax;
-  const AppSettings = require('../common/models/index').AppSettings;
-  const MANIFESTS = require('../common/manifests/index').MANIFESTS;
-  const PivotApplication = require('./applications/pivot-application/pivot-application').PivotApplication;
+  const { WallTime } = require('chronoshift');
+  const { Ajax } = require('./utils/ajax/ajax');
+  const { AppSettings, Timekeeper } = require('../common/models/index');
+  const { MANIFESTS } = require('../common/manifests/index');
+  const { PivotApplication } = require('./applications/pivot-application/pivot-application');
```
Comment (lorem--ipsum) [inline src/client/pivot-entry.ts]: 👍

File: src/common/models/data-cube/data-cube.ts [hunk 4]
```diff
@@ -727,37 +705,33 @@ export class DataCube implements Instance<DataCubeValue, DataCubeJS> {
...
 
-  public getMaxTimeDate(): Date {
-    var { refreshRule } = this;
-    if (refreshRule.isFixed()) return refreshRule.time;
-
-    // refreshRule is query or realtime
-    var { maxTime } = this;
-    if (!maxTime) return null;
-    return second.ceil(maxTime.time, Timezone.UTC);
+  public getMaxTime(timekeeper: Timekeeper): Date {
```
Comment (lorem--ipsum) [inline src/common/models/data-cube/data-cube.ts]: Really cool

File: src/common/models/filter-clause/filter-clause.ts (hunk 2 above)
Comment (lorem--ipsum) [inline src/common/models/filter-clause/filter-clause.ts]: I'd say the two produce the same result

File: src/common/utils/time-monitor/time-monitor.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  constructor(logger: Logger) {
+    this.logger = logger;
+    this.checks = {};
+    this.regularCheckInterval = 60000;
+    this.specialCheckInterval = 60000;
+    this.timekeeper = Timekeeper.EMPTY;
+    setInterval(this.doChecks.bind(this), 1000);
+  }
+
+  addCheck(name: string, check: Check): this {
```
Comment (lorem--ipsum) [inline src/common/utils/time-monitor/time-monitor.ts]: This should be called `addOrUpdateCheck` IMHO

**Comment by lorem--ipsum** [comment]
That's a great improvement ! I'm down for merging it so @evasomething feel free to do so when you want. A non-blocking comment though, I'm not a huge fan of the case for `Timekeeper`. I'd have had written it `TimeKeeper`... Thoughts ?

**Comment by evasomething** [comment]
💧

**Pull request merged**